File: client/src/instruction/configUpdateWorkerFee.ts
import { BN, Program } from "@project-serum/anchor";
import { TransactionInstruction } from "@solana/web3.js";
import { Cronos } from "../idl";
import { Account } from "../account";

export type ConfigUpdateWorkerFeeArgs = {
  newWorkerFee: BN;
};

export class ConfigUpdateWorkerFee {
  private account: Account;
  private cronos: Program<Cronos>;

  constructor(account: Account, cronos: Program<Cronos>) {
    this.account = account;
    this.cronos = cronos;
  }

  public async configUpdateWorkerFee({
    newWorkerFee,
  }: ConfigUpdateWorkerFeeArgs): Promise<TransactionInstruction> {
    const configPDA = await this.account.config.pda();
    const configData = await this.account.config.data(configPDA.address);
    return this.cronos.instruction.configUpdateWorkerFee(newWorkerFee, {
      accounts: {
        admin: configData.admin,
        config: configPDA.address,
      },
    });
  }
}

File: src/instruction/config_update_worker_fee.rs
use anchor_client::anchor_lang::InstructionData;
use solana_program::{
    instruction::{AccountMeta, Instruction},
    pubkey::Pubkey,
};

pub fn config_update_worker_fee(admin: Pubkey, config: Pubkey, new_worker_fee: u64) -> Instruction {
    Instruction {
        program_id: cronos_program::ID,
        accounts: vec![
            AccountMeta::new(admin, true),
            AccountMeta::new(config, false),
        ],
        data: cronos_program::instruction::ConfigUpdateWorkerFee { new_worker_fee }.data(),
    }
}
